modules/foreign-state-oracle/src/lib.rs

```diff
@@ -0,0 +1,277 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Foreign State Oracle.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{
+	dispatch::{Dispatchable, GetDispatchInfo, PostDispatchInfo},
+	pallet_prelude::*,
+	traits::{Currency, ExistenceRequirement, NamedReservableCurrency},
+	transactional, PalletId,
+};
+use frame_system::pallet_prelude::*;
+use module_support::ForeignChainStateQuery;
+use primitives::{Balance, ReserveIdentifier};
+use sp_runtime::{
+	traits::{AccountIdConversion, BlockNumberProvider, One, Saturating, Scale},
+	ArithmeticError,
+};
+use sp_std::{prelude::Vec, vec};
+
+mod mock;
+mod tests;
+
+pub use module::*;
+
+// Unique Identifier for each query
+pub type QueryIndex = u64;
+
+// Origin with arbitrary bytes included
+#[derive(PartialEq, Eq, Clone, RuntimeDebug, Encode, Decode, TypeInfo)]
+pub struct RawOrigin {
+	data: Vec<u8>,
+}
+
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub struct VerifiableCall<Call, BlockNumber, Balance> {
+	// Call to be dispatched by oracle
+	dispatchable_call: Call,
+	// Blocknumber at which call will be expired
+	expiry: BlockNumber,
+	// Reward available for responding to this query
+	oracle_reward: Balance,
+}
+
+#[frame_support::pallet]
+pub mod module {
+	use super::*;
+
+	pub type VerifiableCallOf<T> =
+		VerifiableCall<<T as Config>::VerifiableTask, <T as frame_system::Config>::BlockNumber, Balance>;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// The outer origin type.
+		type Origin: From<RawOrigin>;
+
+		/// Currency for query payments
+		type Currency: NamedReservableCurrency<
+			Self::AccountId,
+			Balance = Balance,
+			ReserveIdentifier = ReserveIdentifier,
+		>;
+
+		/// The foreign state oracle module id, keeps expired queries deposits
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// Fee to be paid to oracles for servicing query
+		#[pallet::constant]
+		type QueryFee: Get<Balance>;
+
+		/// Fee for cancelling query
+		#[pallet::constant]
+		type CancelFee: Get<Balance>;
+
+		/// Timeout for query requests
+		#[pallet::constant]
+		type QueryDuration: Get<Self::BlockNumber>;
+
+		/// Dispatchable task that needs to be verified by oracle for dispatch
+		type VerifiableTask: Parameter
```

Comment:
This is really a dispatchable call. Calling it Task is a bit confusing.

Reply:
Renamed to DispatchableCall